### PR TITLE
Support URLs in the Notes Tab

### PR DESCRIPTION
**Pull request by pHiney**
Fixes allowing URLs in Notes.

### Description of the problem being solved:
Allows a way for build creators to supply URLs to the atlas tree in their builds, offering an alternative to having to support the Atlas Tree.

### Steps taken to verify a working solution:
- Open a build, type some text in the Notes tab.
1. Double Click some text and observe the browser is not opened.
- Type or paste in a URL.
EG: http://reaper.fm
- Double Click the URL and observe the browser opens with the url.
- Type or paste in c:\Windows\notepad.exe .
- Double Click the notepad text and observe the browser is not opened, nor does notepad.

### Some Notes that showcases this PR:
<pre>Notes

Stuff

Link https://tinyurl.com/yp22t3tu
Link http://reaper.fm
Link https://tinyurl.com/yp22t3tu3
Link https://www.lua.org/pil/20.1.html
Link https://www.google.com/search?q=path+of+exile+righteous+fire+supports&client=firefox-b-d&sxsrf=ALiCzsY9NqKBUO9_12uCH8HAWrOyOv6QsA%3A1654593067138&ei=KxafYuqJCLH04-EPrJGxgAg&oq=path+of+Exile+&sclient=gws-wiz
Executable c:\Windows\notepad.exe

More Stuff      

/ ? ://?
.
https://www.youtube.com/watch?v=fpELYXYTGeY</pre>

### New descriptive text
image

edit: 31/05/2022
Merge current DEV and solve merge conflicts

### CODE REVIEW

File: src/Classes/NotesTab.lua [hunk 1]
```diff
@@ -38,6 +42,7 @@ Below are some common color codes PoB uses:	]]
 	self.controls.edit.height = function()
 		return self.height - 112
 	end
+	self.controls.edit.allowURLs = true
```
Comment (Wires77) [inline src/Classes/NotesTab.lua:45]: Is this used anywhere?  Remove it if not.

File: src/Classes/NotesTab.lua [hunk 2]
```diff
@@ -14,8 +14,12 @@ local NotesTabClass = newClass("NotesTab", "ControlHost", "Control", function(se
 	self.lastContent = ""
 	self.showColorCodes = false
 
-	local notesDesc = [[^7You can use Ctrl +/- (or Ctrl+Scroll) to zoom in and out and Ctrl+0 to reset.
-This field also supports different colors.  Using the caret symbol (^) followed by a Hex code or a number (0-9) will set the color.
+	-- local DescriptionText = [[^7This field supports URLs and different colors. URLs can accessed by first clicking on the url and selecting either Ctrl-RightClick or Ctrl-u.
+	-- Colors can be used by using the caret symbol (^) followed by a Hex code or a number (0-9) will set the color. Below are some common color codes PoB uses:	]]
+	-- self.controls.colorDoc = new("LabelControl", {"TOPLEFT",self,"TOPLEFT"}, 8, 8, 150, 16, DescriptionText)
+	-- self.controls.normal = new("ButtonControl", {"TOPLEFT",self.controls.colorDoc,"TOPLEFT"}, 0, 32, 100, 18, colorCodes.NORMAL.."NORMAL", function() self:SetColor(colorCodes.NORMAL) end)
```
Comment (Wires77) [inline src/Classes/NotesTab.lua:20]: Remove this commented out code

**Review by Wires77 — CHANGES_REQUESTED** [review]
We don't need two different hotkeys for opening a URL.  In addition, I think a double click to open a link is a lot more intuitive for users than Ctrl+Rightclick or Ctrl+u.  You could structure the logic similar to the existing double click functionality, but have the caret extend all the way until it hits whitespace on either side, then check if that's a URL instead of having to scan the entire text for a URL.

**Comment by Nostrademous** [comment]
@Wires77 @pHiney just a note - this allows URL's to anything in Notes, not just Atlas Passive Tree. We have to decide if it's our responsibility to restrict to specific domains only to prevent scams / malware / inappropriate materials to be potential spread through URL's enabled in Notes.

In my Trader PR I only allow very specific domains (like PoE Trade and poe.ninja) to be valid URLs.

File: src/Classes/NotesTab.lua [hunk 3]
```diff
@@ -95,6 +116,19 @@ function NotesTabClass:Draw(viewPort, inputEvents)
 			elseif event.key == "y" and IsKeyDown("CTRL") then
 				self.controls.edit:Redo()
 			end
+		elseif event.type == "KeyUp" then
```
[the remark this answers is not in the document]
Comment (pHiney) [reply src/Classes/NotesTab.lua:119]: This is described in the Notes above. To reduce multiple spawning of urls

"Notes

    I moved paste in EditControl to KeyUp
    As the above process was triggering multiple browser tabs for each right click and moving the code for spawning the browser to KeyUp would trigger the paste procedure as well."

As for using KeyDown. That is bad programming practice. The only thing on key down should be what needs to be (eg, detecting and starting drag). 
By keying things to key down you get multiple repeats of code for no reason.

File: src/Classes/NotesTab.lua [hunk 4]
```diff
@@ -95,6 +116,19 @@ function NotesTabClass:Draw(viewPort, inputEvents)
...
+			local ctrl = IsKeyDown("CTRL")
+			if ctrl and ( event.key == "RIGHTBUTTON" or event.key == "u" ) then
+				caret = self.controls.edit.caret
+				if FindURLs() then
+					for _, entry in ipairs(urlList) do
+						if caret >= entry[1] and caret < entry[2] then
+							os.execute('c:/windows/explorer.exe "'..entry[3]..'"')
```
[the remark this answers is not in the document]
Comment (pHiney) [reply src/Classes/NotesTab.lua:126]: I'm sure i tried that first and it didn't work. oh well. Does now.

**Pull request closed without merging**